### PR TITLE
Re-Enabling Cross Compilation for Legion

### PR DESCRIPTION
**Pull request by rspavel**
This resolves a recently discovered issue where, due to the way gasnet is deployed in the current `legion` spackage, cross compilation and spack dependencies are not fully respected and this manifests as problematic on certain platforms.

To support these legacy dependents of legion and gasnet, the previous version of the rewritten gasnet spackage has been restored as `gasnet-legacy` so as to not interfere with use cases of the current `gasnet` spackage. This allows spack to fully deploy `gasnet-legacy` for builds of `legion` that require it and reenables support for certain platforms.

This has the additional bonus of allowing for older `legion` builds and reenabling support for spackages that disabled their legion variants as a result of the rewrites, but that will likely require significantly more effort and is being left for a later PR.

### CODE REVIEW

**Review by bonachea — CHANGES_REQUESTED** [review]
We, the GASNet library developers, have some concerns about this PR, which restores a deeply flawed Spackage for GASNet that we previously rewrote/replaced. We’d like some time to discuss the motivation and use cases of this PR in more depth with both the Legion team and the PR authors to reach consensus on the best solution.

CC: @PHHargrove @pmccormick @streichler

**Comment by rspavel** [comment]
@bonachea I am very cool with making tweaks that better adhere to the preferred gasnet methodology if you have any suggestions. 

But, as we have legacy apps and users, we do need the ability to recreate those results. And the current model doesn't follow spack methodologies and results in issues if a dependent tries to cross compile among other issues.

That is why we specifically set this as a `legacy` spackage and marked every single version as `deprecated`. Users are NOT supposed to use this but... some will need to. And the intent is that, going forward, no other application would use `gasnet-legacy` and would instead use `gasnet` in the appropriate way. But we can't just throw away all existing work and will need a way to demonstrate progress by comparing against older versions on platforms where the front end and compute nodes may not be identical.

**Comment by PHHargrove** [comment]
@rspavel, On Tuesday @bonachea and I met with @pmccormick and we now believe that we understand how to address your problems (_as we understand them_) within the `legion` spack package without the need for this PR.  

Regarding "as we understand them", above, I believe Pat will be following up to confirm he is chasing the correct issues.

**Comment by rspavel** [comment]
> @rspavel, On Tuesday @bonachea and I met with @pmccormick and we now believe that we understand how to address your problems (_as we understand them_) within the `legion` spack package without the need for this PR.
> 
> Regarding "as we understand them", above, I believe Pat will be following up to confirm he is chasing the correct issues.

That still doesn't really allow any version of legion from prior to the rewrite (actually upcoming rewrite) to be built but we already added conflicts to those spackages and probably can find a way to rewrite this fix yet again if we ever have to reproduce results

Beyond that: The easiest way to test the current known issues is: Go on a heterogeneous machine with a particularly restrictive firewall and a system MPI installation. For example, a machine with Haswsell and Broadwell nodes. Build on Broadwell with a generic x86 target and a spack built MPI. Switch to a haswell node and:

1. Check to make sure the system MPI is not being used by legion/gasnet
2. Check to see if the generic x86 target legion can run on the Haswell 

That should check if the rules and environment of spack are being respected.

But yeah, let's keep this PR open until there is one with the legion preferred fix that we can test against.

**Comment by pmccormick** [comment]
@rspavel: Can you give us an example command line that you are using, including the mpi variant, etc.  We suspect a few issues are lurking but something specific from your use case would be helpful in debugging and adding any necessary parameters for better control of the details.   Thanks.

**Comment by rspavel** [comment]
> @rspavel: Can you give us an example command line that you are using, including the mpi variant, etc. We suspect a few issues are lurking but something specific from your use case would be helpful in debugging and adding any necessary parameters for better control of the details. Thanks.

Assuming a machine that uses slurm and has one partition `broadwell` and one partition `haswell` that behave as one would expect. I used flecsi to provide a rapid way to test but:


This works because it has spack build gasnet in a way that respects the specified target
```bash
salloc -N 1 - p broadwell
spack env create -d . && spack env activate -d .
 spack add flecsi@2.1.0%gcc@9.3.0~cuda~external_cinch+kokkos~shared backend=legion ^legion+gasnet_legacy ^mpich+slurm %gcc@9.3.0 ^kokkos@3.3.01 ^boost@1.70.0 arch=linux-centos7-x86_64 
spack concretize && spack install
spack env loads --dependencies -x flecsi
#De-allocate allocation
salloc -N 1 -p haswell
source loads
#Build and test flecsi
```

and this would, presumably, use the arch of the node you are on when building gasnet and throw up a lot of illegal instructions if you run it on an older intel

```bash
salloc -N 1 - p broadwell
spack env create -d . && spack env activate -d .
 spack add flecsi@2.1.0%gcc@9.3.0~cuda~external_cinch+kokkos~shared backend=legion ^legion~gasnet_legacy ^mpich+slurm %gcc@9.3.0 ^kokkos@3.3.01 ^boost@1.70.0 arch=linux-centos7-x86_64 
spack concretize && spack install
spack env loads --dependencies -x flecsi
#De-allocate allocation
salloc -N 1 -p haswell
source loads
#Build and test flecsi
```

I don't have an easy tester for whether system mpi (and other dependencies) are pulled in but have helped a few people debug issues related to that. Harder to make a reproducer for something like that.


Also, considering that an upcoming fix would only apply to versions of legion after that, is there any objection to just merging this PR in to support legacy legion versions and to allow us to reproduce results for versions of legion that are not capable of deploying their own gasnet at all? We can add a conflict once legion is able to successfully deploy its own gasnet but this is really solving two different problems at this point. Because this PR supports all current build system legions (near as I can tell) and it won't be hard to provide support for previous build system legion as well.


So are there any actual objections to this as a fix to support older versions of legion?

**Comment by bonachea** [comment]
@rspavel Thanks for the detailed example, although I suspect you made a copy/paste error in the second example (which looks identical to the first?). @pmccormick will likely followup on this.

> Also, considering that an upcoming fix would only apply to versions of legion after that, 

I'm not sure this statement is true. IIUC if @pmccormick applies fixes to address your concerns within the Legion spackage that don't change the Legion build system, then it might "just work" for _some_ older Legion versions. 

**Can you please clarify the oldest Legion, GASNet and Flecsi versions you _actually_ need to support in testing?**

@rspavel wrote:
> So are there any actual objections to this as a fix to support older versions of legion?

Yes, as I indicated in my first post we **the GASNet maintainers have _multiple_ technical objections** to restoring the old, deeply flawed GASNet spackage, which we completely rewrote/replaced specifically to remedy those problems. I have not yet posted in detail on those objections because I'd much rather see us pursue a solution by "working downwards" from Pat's Legion spackage, rather than starting from the old/terrible GASNet Spackage and "working upwards" to fix all the problems we'd need to address to reach consensus (on what would still fundamentally be a "competing" GASNet spackage).

**Comment by rspavel** [comment]
> @rspavel Thanks for the detailed example, although I suspect you made a copy/paste error in the second example (which looks identical to the first?). @pmccormick will likely followup on this.
> 
> > Also, considering that an upcoming fix would only apply to versions of legion after that,
> 
> I'm not sure this statement is true. IIUC if @pmccormick applies fixes to address your concerns within the Legion spackage that don't change the Legion build system, then it might "just work" for _some_ older Legion versions.
> 
> **Can you please clarify the oldest Legion, GASNet and Flecsi versions you _actually_ need to support in testing?**
> 
> @rspavel wrote:
> 
> > So are there any actual objections to this as a fix to support older versions of legion?
> 
> Yes, as I indicated in my first post we **the GASNet maintainers have _multiple_ technical objections** to restoring the old, deeply flawed GASNet spackage, which we completely rewrote/replaced specifically to remedy those problems. I have not yet posted in detail on those objections because I'd much rather see us pursue a solution by "working downwards" from Pat's Legion spackage, rather than starting from the old/terrible GASNet Spackage and "working upwards" to fix all the problems we'd need to address to reach consensus (on what would still fundamentally be a "competing" GASNet spackage).

As someone who is not involved in these private meetings but is expected to help reproduce legacy results to demonstrate performance improvements over time, would you be willing to list those technical objections that will prevent this work? The purpose of a `foo-legacy` spackage, as well as the indication that all of these versions are deprecated, is to indicate that they should not be used. But, unfortunately, they still need to be available when reproducing older builds to monitor performance changes and capability changes. That is just the nature of the beast.

I am perfectly happy to try to address some of those changes. But if the only objection is nebulous and non-specific then I don't actually see any technical reason the two solutions can't coexist. The updated legion spackage will support all use cases going forward. The legacy variant in the legion spackage and the `gasnet-legacy` spackage will allow us to produce results from before some time last year or whenever the changeover happened.

I appreciate that gasnet was not used correctly in the past. Unfortunately, ECP and LANL ASC work was built on that and we now have a solution that allows us to make forward progress without throwing all of that out (well, pulling that out of the bin since it was already thrown out during the rewrite).

**Comment by bonachea** [comment]
Pull request #27860 provides a surgical fix to the primary "cross-architecture compilation" defect described in this PR.

My understanding is that @pmccormick is continuing to work with @scothalverson and @rspavel on addressing other requests raised in this PR through alternative further improvements to the Legion spackage that should be coming soon..

**Comment by rspavel** [comment]
> Pull request #27860 provides a surgical fix to the primary "cross-architecture compilation" defect described in this PR.
> 
> My understanding is that @pmccormick is continuing to work with @scothalverson and @rspavel on addressing other requests raised in this PR through alternative further improvements to the Legion spackage that should be coming soon..

We are currently in discussions as to whether that PR fixes all issues or only a subset of the ones this PR was meant to fix

**Pull request closed without merging**
